streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java

```diff
@@ -124,6 +125,28 @@ public void shouldNotOptimizeWhenAThroughOperationIsDone() {
 
     }
 
+    @Test
+    public void shouldOptimizeSeveralMergeNodesWithCommonKeyChangingParent() {
+        final StreamsBuilder streamsBuilder = new StreamsBuilder();
+        final KStream<Integer, Integer> parentStream = streamsBuilder.stream("input_topic", Consumed.with(Serdes.Integer(), Serdes.Integer()))
+            .selectKey(Integer::sum);
+
+        final KStream<Integer, Integer> childStream1 = parentStream.mapValues(v -> v + 1);
+        final KStream<Integer, Integer> childStream2 = parentStream.mapValues(v -> v + 2);
+        final KStream<Integer, Integer> childStream3 = parentStream.mapValues(v -> v + 3);
```

Comment:
While this is the example from the ticket, the issue is not related to branching out as done here. Should we rewrite/simplify the test and just create 3 `KStream sX = streams.builder()` and later add one single key-changing operation? -> `s1.selectKey().merge(s2).merge(s3).`